### ISSUE TITLE
Register service using interface as type / Typing for named services

### ISSUE THREAD

**Issue by slavafomin** [pleerock/typedi#16]
Another issue, I've found when migrating from Angular Injector is that there is no way to register service using interface as a type.

The problem is that interfaces are design-time only artifacts, which are deleted from compiled code altogether. So there is no way to reference the service by it's interface name at runtime.

Angular approached this problem by providing a special class called `InjectionToken`. The workflow is the following: first, you create an instance of the injection token: `export const DI_SEQUELIZE = new InjectionToken<Sequelize>('sequelize');`, then you use it to register service with the container. By using such hint, typescript can correctly track types of the instances returned from the container.

Right now, if I register a named service and then get it, the return type is not inferred, so I have to manually cast it to a correct type, which makes code harder to read and maintain.

@pleerock What could be done about this?

**Comment by pleerock** [comment]
We can add same functionality:

```ts
export const SequelizeToken = new Token<Sequelize>("badorm");
```

then simply:

```ts
const badOrmInstance = Container.get(SequelizeToken);
```

in container.get a check:

```ts
if (name instanceof Token) {
   // ...
}
```

Need to think about naming of the `Token`. I totally dislike `InjectionToken`. Simply `Token` is okay but would be good to have something better.

**Comment by pleerock** [comment]
probably it can be even simply `export const SequelizeToken = new Token<Sequelize>();` - there is no need in label anymore, because `SequelizeToken` itself is an unique identifier.

**Comment by slavafomin** [comment]
> "badorm"

:smile:

Off-topic: I'm also migrating from Sequelize at the moment, btw, could you elaborate on why do you consider it as bad?

**Comment by pleerock** [comment]
it was a joke. Sequelize is mature and stable, much more then typeorm at this moment. Its just personal preference. I prefer DataMapper over ActiveRecord and I don't like how javascript-orm looks and feels like.

**Comment by pleerock** [comment]
do you have in plans to work on this issue? - so we can release everything together in next 0.5.0 version.

**Comment by pleerock** [comment]
fixed in #18
